emacs/config.org
#+AUTHOR: Zheng Zhicheng(ZZC)
#+DESCRIPTION: Just another Emacs config from scratch
#+STARTUP: overview
#+PROPERTY: header-args:emacs-lisp :tangle ~/dotconfig/emacs/init.el

* Startup

#+begin_src emacs-lisp
  ;; -*- lexical-binding: t; -*-
  ;;show show errors
  (setq warning-minimum-level :error)
  (setq custom-file (expand-file-name "custom.el" user-emacs-directory))
  ;;disable bell
  (setq visible-bell 1)
  ;; The default is 800 kilobytes.  Measured in bytes.
  (setq gc-cons-threshold (* 50 1000 1000))

  ;; Profile emacs startup
  (add-hook 'emacs-startup-hook
  	  (lambda ()
  	    (message "*** Emacs loaded in %s seconds with %d garbage collections."
  		     (emacs-init-time "%.2f")
  		     gcs-done)))

  (defvar native-comp-deferred-compilation-deny-list nil)

#+end_src


* Package Management
** Straight
#+begin_src emacs-lisp
  (defvar bootstrap-version)
  ;; ?? Emacs 29 ??? native-compile ??????? bug
  (unless (version<= emacs-version "28.2")
    (setq straight-repository-branch "develop"))
  (setq straight-check-for-modifications '(check-on-save find-when-checking))

  (let ((bootstrap-file
         (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
        (bootstrap-version 6))
    (unless (file-exists-p bootstrap-file)
      (with-current-buffer
          (url-retrieve-synchronously
           "https://raw.githubusercontent.com/radian-software/straight.el/develop/install.el"
           'silent 'inhibit-cookies)
        (goto-char (point-max))
        (eval-print-last-sexp)))
    (load bootstrap-file nil 'nomessage))
  (straight-use-package 'use-package)
  (setq straight-use-package-by-default t)
  (setq use-package-always-ensure t)
#+end_src

** Package path

#+begin_src emacs-lisp
  ;; make sure shell PATH is same as emacs PATH 
  (use-package exec-path-from-shell
    :config
    (setq shell-file-name "/bin/bash")
    (setq exec-path-from-shell-arguments '("-l"))
    (when (memq window-system '(mac ns x))
      (exec-path-from-shell-initialize)))
#+end_src

** Starup Profiling
#+BEGIN_SRC emacs-lisp
  (use-package esup
    :config
    (setq esup-depth 0))
#+END_SRC


* Path
#+BEGIN_SRC emacs-lisp
  (setq org_notes_dir "~/Documents/org/notes" ; org notes location
        zot_bib "~/Nutstore/1/Nutstore/Zotero-Library/Main.bib"; Zotero .bib 文件
        zot_pdf "~/Nutstore/1/Nutstore/Zotero-Library" ; Zotero 同步文件
        org_notes "~/Documents/org/notes/ref/") ; org-roam 文献笔记目录

  (unless (file-exists-p org_notes_dir) (setq org_notes_dir nil))
  (unless (file-exists-p zot_bib) (setq zot_bib nil))
  (unless (file-exists-p zot_pdf) (setq zot_pdf nil))
  (unless (file-exists-p org_notes) (setq org_notes nil)) ; 防止文件不存在报错
#+END_SRC


* System related
** Detect System
#+begin_src emacs-lisp
  (setq my/is-windows (eq system-type 'windows-nt)) ; Windows 
  (setq my/is-linux (eq system-type 'gnu/linux)) ; Linux
  (setq my/is-mac (eq system-type 'darwin)) ; mac
  (setq my/is-WSL
        (if (and (eq system-type 'gnu/linux)
                 (string-match-p "Microsoft" (shell-command-to-string "uname -r")))
            t
          nil)) ; WSL
  (setq my/is-terminal (not window-system)) ;GUI
#+end_src

** WSL

*** Browse url

#+begin_src emacs-lisp
  (when (and (eq system-type 'gnu/linux)
             (string-match
              "Linux.*Microsoft.*Linux"
              (shell-command-to-string "uname -a")))
    (setq
     browse-url-generic-program  "/mnt/c/Windows/System32/cmd.exe"
     browse-url-generic-args     '("/c" "start")
     browse-url-browser-function #'browse-url-generic))

#+end_src

** Misc
#+BEGIN_SRC emacs-lisp
  (use-package no-littering
    :demand t
    :custom
    (auto-save-file-name-transforms `((".*" ,(no-littering-expand-var-file-name "auto-save/") t)))) ; 设置自动保存文件目录
  (use-package recentf
  :after no-littering
  :demand t 
  :custom
  (recentf-exclude '(no-littering-var-directory
                     no-littering-etc-directory)) ; 屏蔽临时文件
  (recentf-max-menu-items 25)
  (recentf-max-saved-items 25)
  :bind ("C-x C-r" . 'recentf-open-files)
  :config
  (recentf-mode 1))
#+END_SRC


* History and save 
#+BEGIN_SRC emacs-lisp
(use-package saveplace
  :defer 1
  :config
    (save-place-mode 1))
(use-package savehist
  :defer 1
  :config (savehist-mode))
(use-package super-save
  :defer 1
  :custom
  (super-save-auto-save-when-idle t)
  :config
  (super-save-mode +1))
(global-auto-revert-mode 1)
(setq global-auto-revert-non-file-buffers nil)
#+END_SRC


* Key Binding
** Which-key 
#+begin_src emacs-lisp
  (use-package which-key
    :init (which-key-mode)
    :diminish which-key-mode
    :config
    (setq which-key-idle-delay 1))
#+end_src

** General
#+begin_src emacs-lisp
  (use-package general
    :config
    (general-create-definer zzc/leader-keys
      :states '(normal visual emacs)
      :keymaps 'override
      :prefix "SPC"
      :global-prefix "C-SPC"))
#+end_src

** EVIL
#+begin_src emacs-lisp
  (use-package evil
    :demand t
    :bind (("<escape>" . keyboard-escape-quit))
    :init
    (setq evil-want-integration t)
    (setq evil-symbol-word-search t)
    (setq evil-want-keybinding nil)
    (setq evil-want-C-u-scroll t)
    (setq evil-want-C-i-jump t)
    :config
    (evil-mode 1)
    ;; (define-key evil-insert-state-map (kbd "C-g") 'evil-normal-state)
    ;; Use visual line motions even outside of visual-line-mode buffers
    (evil-global-set-key 'motion "j" 'evil-next-visual-line)
    (evil-global-set-key 'motion "k" 'evil-previous-visual-line)
    (evil-set-initial-state 'messages-buffer-mode 'normal)
    (evil-set-initial-state 'dashboard-mode 'normal))
    
  (use-package evil-collection
    :after evil
    :config
    (evil-collection-init))
  (use-package evil-commentary
    :after evil
    :init (evil-commentary-mode))
  (use-package evil-surround
    :after evil
    :config
    (global-evil-surround-mode 1))
  (use-package undo-tree
    :after evil
    :diminish
    :config
    (evil-set-undo-system 'undo-tree)
    (global-undo-tree-mode 1))
  (setq undo-tree-history-directory-alist '(("." . "~/.emacs.d/undo")))
  (use-package evil-escape
    :init (evil-escape-mode)
    :after evil
    :config
    (setq evil-escape-key-sequence "jj")
    (setq evil-escape-delay 0.3))
  (use-package evil-org
    :after org
    :hook (org-mode . evil-org-mode)
    :config
    (require 'evil-org-agenda)
    (evil-org-agenda-set-keys)
    (general-define-key
     :states '(normal motion)
     :keymaps 'org-mode-map
     "zd" 'org-fold-hide-drawer-toggle))

  ;; save file very time after quit insert mode
  (add-hook 'evil-insert-state-exit-hook
            (lambda ()
              (call-interactively #'save-buffer)))

#+end_src


* Buffer & Window
**  help buffer
#+BEGIN_SRC emacs-lisp
  (setq
   display-buffer-alist
   '(("^\\*[Hh]elp"                            ;正则匹配 buffer name
      (display-buffer-reuse-window
     ;入口函数，一个个调用直到有返回值，参数是：1.buffer 2.剩下的这些 alist
       display-buffer-in-side-window)
      (side . right)                        ;参数 alist 从这里开始。这个 side 会被 display-buffer-in-side-window 使用
      (window-width . 0.5)                     ;emacs 会自动把这个设置到 window-parameter 里
      (window-height . 0.33)                   ;同上
      (slot . 1)                               ;这个会被 display-buffer-in-side-window 使用，控制 window 位置
      (reusable-frames . visible)              ;这个参数看第三个链接的 display-buffer
      (haha . whatever)                        ;当然随你放什么
      (window-parameters                       ;emacs 26 及以上会自动把下面的设置到 window-parameter 里
       (select . t)                            ;自定义的 param
       (quit . t)                              ;同上
       (popup . t)                             ;同上
       (mode-line-format . none)               ;emacs version > 25， none 会隐藏 mode line，nil 会显示...
       (no-other-window . t)                   ;随你设置其他的 window-parameter，看文档
       ))))

#+END_SRC

** Focus on new split
#+BEGIN_SRC emacs-lisp
  (defun split-window-right-and-focus ()
    "Split the window vertically and move focus to the new one."
    (interactive)
    (split-window-right)
    (other-window 1))

  (defun split-window-below-and-focus ()
    "Split the window horizontally and move focus to the new one."
    (interactive)
    (split-window-below)
    (other-window 1))

  (defadvice split-window-right (after split-window-right-and-focus activate)
    (other-window 1))

  (defadvice split-window-below (after split-window-below-and-focus activate)
    (other-window 1))
#+END_SRC

** Project

#+BEGIN_SRC emacs-lisp
  (use-package project
    ;; Cannot use :hook because 'project-find-functions does not end in -hook
    ;; Cannot use :init (must use :config) because otherwise
    ;; project-find-functions is not yet initialized.
    :config
    (setq project-vc-extra-root-markers '(".project" "*.csproj")))

#+END_SRC

** Perspective

#+BEGIN_SRC emacs-lisp
  (use-package perspective
    :bind
    ("C-x C-b" . persp-list-buffers)         ; or use a nicer switcher, see below
  ;; :custom
    ;; (persp-mode-prefix-key (kbd "SPC p"))  ; pick your own prefix key here
    :config
    (setq persp-state-default-file (expand-file-name ".persp-save" user-emacs-directory))
    ;; Save perspectives without confirmation

    (defun my/persp-state-save-silent ()
    "Save perspective state without confirmation."
    (let ((persp-state-save-behavior nil)) ; Prevent prompting
      (persp-state-save persp-state-default-file)))

    ;; Load perspectives without confirmation
    (defun my/persp-state-load-silent ()
      "Load perspective state without confirmation."
      (when (file-exists-p persp-state-default-file)
        (persp-state-load persp-state-default-file)))

    ;; Automatically save perspectives when Emacs quits
    (add-hook 'kill-emacs-hook #'my/persp-state-save-silent)
    ;; Automatically load perspectives at startup
    (add-hook 'emacs-startup-hook #'my/persp-state-load-silent)
  (zzc/leader-keys
    "p" '(:keymap perspective-map :which-key "perspective")
    :package 'perspective)
    :init
    (persp-mode))

#+END_SRC

** Buffers
*** Never save scratch 
#+BEGIN_SRC emacs-lisp
(defun my-scratch-buffer-no-save ()
  "Prevent any buffer with *scratch* in its name from being marked as modified."
  (interactive)
  (dolist (buf (buffer-list))
    (when (string-match-p "\\*scratch\\*" (buffer-name buf))
      (with-current-buffer buf
        (set-buffer-modified-p nil)))))
(add-hook 'after-change-functions (lambda (&rest _) (my-scratch-buffer-no-save)))
#+END_SRC

*** Key mapping

#+begin_src emacs-lisp
  (zzc/leader-keys
    "b"  '(:ignore t :which-key "buffer")
    "bp"  '(switch-to-prev-buffer :which-key "previous buffer")
    "bn"  '(switch-to-next-buffer :which-key "next buffer")
    "bb"  '(switch-to-buffer :which-key "list buffers")
    "bB"  '(ibuffer-list-buffers :which-key "list ibuffers")
    "bd"  '(kill-current-buffer :which-key "kill current buffer")
    "bs"  '(save-buffer :which-key "save buffer")
    )
#+end_src

** Winner mode

#+begin_src emacs-lisp
  (winner-mode 1)
  (global-set-key (kbd "C-c u") 'winner-undo)
  (global-set-key (kbd "C-c r") 'winner-redo)

#+end_src

** Windows number
#+BEGIN_SRC emacs-lisp
  (use-package winum
    :ensure t
    :config
    (winum-mode))
#+END_SRC

** Maxmize window toggle

#+BEGIN_SRC emacs-lisp
  (defvar toggle-one-window-window-configuration nil
    "The window configuration use for `toggle-one-window'.")
  (defun toggle-one-window ()
    "Toggle between window layout and one window."
    (interactive)
    (if (equal (length (cl-remove-if #'window-dedicated-p (window-list))) 1)
        (if toggle-one-window-window-configuration
            (progn
              (set-window-configuration toggle-one-window-window-configuration)
              (setq toggle-one-window-window-configuration nil))
          (message "No other windows exist."))
      (setq toggle-one-window-window-configuration (current-window-configuration))
      (delete-other-windows)))
  (general-define-key
   :prefix "C-c"
   "m" 'toggle-one-window)

#+END_SRC


* File
** Bookmarks
#+begin_src emacs-lisp
    ;; save bookmark on change
    (setq bookmark-save-flag 1)
    (require 'bookmark)
    ;; set bookmark file to sync across difference device
    (setq bookmark-default-file "~/.config/emacs/bookmarks")
    (zzc/leader-keys
      "bm"  '(:ignore t :which-key "bookmark")
      "bmm"  '(bookmark-set :which-key "Add current file/dir to bookmark")
      "bml"  '(list-bookmarks :which-key "Open Bookmark List"))

#+end_src

** Open specific file 

#+begin_src emacs-lisp
  (zzc/leader-keys
    "="  '(:ignore t :which-key "open")
    "=h" '((lambda () (interactive) (find-file "~/Documents/notes/20241004160632-habit_tracking.org")) :which-key "open habit.org")
    "=c" '((lambda () (interactive) (find-file "~/.config/emacs/config.org")) :which-key "open config file"))
#+end_src

** Key mapping
#+begin_src emacs-lisp
  (zzc/leader-keys
    "."  '(find-file :which-key "find file")
    )
#+end_src

** Treemacs
#+BEGIN_SRC emacs-lisp
  (use-package treemacs
    :ensure t
    :defer t
    :init
    (with-eval-after-load 'winum
      (define-key winum-keymap (kbd "M-0") #'treemacs-select-window))
    :config
    (progn
      (setq treemacs-collapse-dirs                   (if treemacs-python-executable 3 0)
            treemacs-deferred-git-apply-delay        0.5
            treemacs-directory-name-transformer      #'identity
            treemacs-display-in-side-window          t
            treemacs-eldoc-display                   'simple
            treemacs-file-event-delay                2000
            treemacs-file-extension-regex            treemacs-last-period-regex-value
            treemacs-file-follow-delay               0.2
            treemacs-file-name-transformer           #'identity
            treemacs-follow-after-init               t
            treemacs-expand-after-init               t
            treemacs-find-workspace-method           'find-for-file-or-pick-first
            treemacs-git-command-pipe                ""
            treemacs-goto-tag-strategy               'refetch-index
            treemacs-header-scroll-indicators        '(nil . "^^^^^^")
            treemacs-hide-dot-git-directory          t
            treemacs-indentation                     2
            treemacs-indentation-string              " "
            treemacs-is-never-other-window           nil
            treemacs-max-git-entries                 5000
            treemacs-missing-project-action          'ask
            treemacs-move-files-by-mouse-dragging    t
            treemacs-move-forward-on-expand          nil
            treemacs-no-png-images                   nil
            treemacs-no-delete-other-windows         t
            treemacs-project-follow-cleanup          nil
            treemacs-persist-file                    (expand-file-name ".cache/treemacs-persist" user-emacs-directory)
            treemacs-position                        'left
            treemacs-read-string-input               'from-child-frame
            treemacs-recenter-distance               0.1
            treemacs-recenter-after-file-follow      nil
            treemacs-recenter-after-tag-follow       nil
            treemacs-recenter-after-project-jump     'always
            treemacs-recenter-after-project-expand   'on-distance
            treemacs-litter-directories              '("/node_modules" "/.venv" "/.cask")
            treemacs-project-follow-into-home        nil
            treemacs-show-cursor                     nil
            treemacs-show-hidden-files               t
            treemacs-silent-filewatch                nil
            treemacs-silent-refresh                  nil
            treemacs-sorting                         'alphabetic-asc
            treemacs-select-when-already-in-treemacs 'move-back
            treemacs-space-between-root-nodes        t
            treemacs-tag-follow-cleanup              t
            treemacs-tag-follow-delay                1.5
            treemacs-text-scale                      nil
            treemacs-user-mode-line-format           nil
            treemacs-user-header-line-format         nil
            treemacs-wide-toggle-width               70
            treemacs-width                           35
            treemacs-width-increment                 1
            treemacs-width-is-initially-locked       t
            treemacs-workspace-switch-cleanup        nil)

      ;; The default width and height of the icons is 22 pixels. If you are
      ;; using a Hi-DPI display, uncomment this to double the icon size.
      ;; (treemacs-resize-icons 44)

      (treemacs-follow-mode t)
      (treemacs-filewatch-mode t)
      (treemacs-fringe-indicator-mode 'always)
      (when treemacs-python-executable
        (treemacs-git-commit-diff-mode t))

      (pcase (cons (not (null (executable-find "git")))
                   (not (null treemacs-python-executable)))
        (`(t . t)
         (treemacs-git-mode 'deferred))
        (`(t . _)
         (treemacs-git-mode 'simple)))

      (treemacs-hide-gitignored-files-mode nil))
    :bind
    (:map global-map
          ("M-0"       . treemacs-select-window)
          ("C-x t 1"   . treemacs-delete-other-windows)
          ("C-x t t"   . treemacs)
          ("C-x t d"   . treemacs-select-directory)
          ("C-x t B"   . treemacs-bookmark)
          ("C-x t C-t" . treemacs-find-file)
          ("C-x t M-t" . treemacs-find-tag)))

  (use-package treemacs-evil
    :after (treemacs evil)
    :ensure t)

  (use-package treemacs-icons-dired
    :hook (dired-mode . treemacs-icons-dired-enable-once)
    :ensure t)

  (use-package treemacs-persp ;;treemacs-perspective if you use perspective.el vs. persp-mode

emacs/doom/config.el
;; -*- lexical-binding: t; -*-
(setq org_notes_dir "~/Documents/org/notes" ; org notes location
      zot_bib "~/Nutstore/1/Nutstore/Zotero-Library/Main.bib"; Zotero .bib 文件
      zot_pdf "~/Nutstore/1/Nutstore/Zotero-Library" ; Zotero 同步文件
      org_notes "~/Documents/org/notes/ref/") ; org-roam 文献笔记目录
(setq auth-sources '("~/.authinfo" "~/.authinfo.gpg"))
  (unless (file-exists-p org_notes_dir) (setq org_notes_dir nil))
  (unless (file-exists-p zot_bib) (setq zot_bib nil))
  (unless (file-exists-p zot_pdf) (setq zot_pdf nil))
  (unless (file-exists-p org_notes) (setq org_notes nil)) ; 防止文件不存在报错

(setq my/is-windows (eq system-type 'windows-nt)) ; Windows
(setq my/is-linux (eq system-type 'gnu/linux)) ; Linux
(setq my/is-mac (eq system-type 'darwin)) ; mac
(setq my/is-WSL
      (if (and (eq system-type 'gnu/linux)
               (string-match-p "Microsoft" (shell-command-to-string "uname -r")))
          t
        nil)) ; WSL
(setq my/is-terminal (not window-system)) ;GUI

(when (and (eq system-type 'gnu/linux)
           (string-match
            "Linux.*Microsoft.*Linux"
            (shell-command-to-string "uname -a")))
  (setq
   browse-url-generic-program  "/mnt/c/Windows/System32/cmd.exe"
   browse-url-generic-args     '("/c" "start")
   browse-url-browser-function #'browse-url-generic))

(after! evil
  (setq evil-escape-key-sequence "jj")
  (setq evil-escape-delay 0.2)) ;; Adjust delay if needed

(defun my/resize-window-up ()
  "Increase window height repeatedly."
  (interactive)
  (enlarge-window 1)
  (set-transient-map
   (let ((map (make-sparse-keymap)))
     (define-key map (kbd "<up>") 'my/resize-window-up)
     map)
   t))

(defun my/resize-window-down ()
  "Decrease window height repeatedly."
  (interactive)
  (shrink-window 1)
  (set-transient-map
   (let ((map (make-sparse-keymap)))
     (define-key map (kbd "<down>") 'my/resize-window-down)
     map)
   t))

(defun my/resize-window-right ()
  "Increase window width repeatedly."
  (interactive)
  (enlarge-window-horizontally 1)
  (set-transient-map
   (let ((map (make-sparse-keymap)))
     (define-key map (kbd "<right>") 'my/resize-window-right)
     map)
   t))

(defun my/resize-window-left ()
  "Decrease window width repeatedly."
  (interactive)
  (shrink-window-horizontally 1)
  (set-transient-map
   (let ((map (make-sparse-keymap)))
     (define-key map (kbd "<left>") 'my/resize-window-left)
     map)
   t))

;; Bind these to SPC + arrow keys
(map! :leader
      :desc "Resize window up"    "<up>"    #'my/resize-window-up
      :desc "Resize window down"  "<down>"  #'my/resize-window-down
      :desc "Resize window right" "<right>" #'my/resize-window-right
      :desc "Resize window left"  "<left>"  #'my/resize-window-left)

(defvar toggle-one-window-window-configuration nil
  "The window configuration use for `toggle-one-window'.")
(defun toggle-one-window ()
  "Toggle between window layout and one window."
  (interactive)
  (if (equal (length (cl-remove-if #'window-dedicated-p (window-list))) 1)
      (if toggle-one-window-window-configuration
          (progn
            (set-window-configuration toggle-one-window-window-configuration)
            (setq toggle-one-window-window-configuration nil))
        (message "No other windows exist."))
    (setq toggle-one-window-window-configuration (current-window-configuration))
    (delete-other-windows)))
(general-define-key
 :prefix "C-c"
 "m" 'toggle-one-window)

;; Enable desktop-save-mode to save sessions
(desktop-save-mode 1)

;; Configure auto-saving of the desktop
(setq desktop-auto-save-timeout (* 15 60)) ; 15 minutes in seconds

;; Ensure desktop saves even if Emacs is idle
(add-hook 'auto-save-hook 'desktop-save-in-desktop-dir)

(setq doom-theme 'doom-gruvbox)
(setq display-line-numbers-type t)
(setq doom-font (font-spec :family "JetBrains Mono" :size 22)
      doom-variable-pitch-font (font-spec :family "Cantarell" :size 15)
      doom-big-font (font-spec :family "JetBrains Mono" :size 34))
(after! doom-themes
  (setq doom-themes-enable-bold t
        doom-themes-enable-italic t))
(custom-set-faces!
  '(font-lock-comment-face :slant italic)
  '(font-lock-keyword-face :slant italic))

(setq doom-modeline-persp-name t) ;; Show workspace name in modeline
(setq doom-modeline-display-default-persp-name t) ;; Display the default workspace name
(defun my/display-all-workspaces ()
  "Display all workspace names and numbers in the mode line."
  (let ((workspaces (persp-names)))
    (if workspaces
        (mapconcat (lambda (ws) (format "[%s]" ws)) workspaces " ")
      "No Workspaces")))

(setq-default mode-line-format
              (append mode-line-format
                      '((:eval (my/display-all-workspaces)))))
(setq doom-modeline-workspace-name (lambda () (my/display-all-workspaces)))

(setq org-directory "~/Documents/org/")

(after! org
  (setq org-agenda-dir "~/Documents/org/jira/")
  (setq org-agenda-files (directory-files-recursively org-agenda-dir "\\.org$"))
  ;; save all org files after change todo
  (defmacro η (fnc)
     "return function that ignores its arguments and invokes fnc."
     `(lambda (&rest _rest)
      (funcall ,fnc)))
  (advice-add 'org-deadline       :after (η #'org-save-all-org-buffers))
  (advice-add 'org-schedule       :after (η #'org-save-all-org-buffers))
  (advice-add 'org-store-log-note :after (η #'org-save-all-org-buffers))
  (advice-add 'org-todo           :after (η #'org-save-all-org-buffers))
  (advice-add 'org-priority       :after (η #'org-save-all-org-buffers)))

(after! org
  (defun my/org-refile-update-targets ()
    "Update `org-refile-targets` to match `org-agenda-files`."
    (setq org-refile-targets
          (mapcar (lambda (file) (cons file '(:maxlevel . 3))) org-agenda-files)))

  ;; Run once on startup
  (my/org-refile-update-targets)

  ;; Update targets whenever `org-agenda-files` changes
  (add-hook! 'org-agenda-mode-hook #'my/org-refile-update-targets)
  (add-hook! 'org-mode-hook #'my/org-refile-update-targets))

(after! org-pomodoro
  (setq org-pomodoro-audio-player "mpv"
    org-pomodoro-ticking-sound-p t
    org-pomodoro-ticking-sound-states '(:pomodoro)
    org-pomodoro-finished-sound-p t
    org-pomodoro-short-break-length 5
    org-pomodoro-finished-sound-args "--volume=50"
    org-pomodoro-long-break-sound-args "--volume=50"
    org-pomodoro-short-break-sound-args "--volume=50"
    org-pomodoro-ticking-sound-args "--volume=60"))

(after! org-jira
  (setq org-jira-working-dir "~/Documents/org/jira/")
  (setq jiralib-url "https://jira.vni.agileci.conti.de")
  (setq jiralib-token
    (cons "Authorization"
      (concat "Bearer " (auth-source-pick-first-password
			 :host "jira.vni.agileci.conti.de"))))
  (setq org-jira-use-status-as-todo nil)
  (setq org-jira-jira-status-to-org-keyword-alist
   '(("Working" . "STRT")
     ("New" . "TODO")
     ("Ready" . "TODO")
     ("Closed" . "DONE")
     ("Verifying" . "STRT"))))

(after! org-anki
  (setq org-anki-default-deck "Mega"))

(setq my/daily-note-filename "%<%Y-%m-%d>.org"
      my/daily-note-header "#+title: %<%Y-%m-%d %a>\n\n[[roam:%<%Y-w%W>]]\n\n[[roam:%<%Y-%B>]]\n\n* Tasks\n** Completed\n** Meeting\n\n* Capture\n** Information\n** Opinions\n** Tools\n** Feelings\n\n* Reflection\n** One thing Good\n** One thing Bad\n** Questions to my self\n*** All the decisions make today, which are by choice, and which are by fear?\n* AI Summary")
(defvar my/org-roam-project-template
  '("p" "project" plain "** TODO %?"
    :if-new (file+head+olp "%<%Y%m%d%H>-${slug}.org"
                           "#+title: ${title}\n\n#+category: ${title}\n#+filetags: Project\n"
                           ("tasks"))))
(defun my/org-roam-filter-by-tag (tag-name)
  (lambda (node)
    (member tag-name (org-roam-node-tags node))))
(defun my/org-roam-list-notes-by-tag (tag-name)
  (mapcar #'org-roam-node-file
    (seq-filter
      (my/org-roam-filter-by-tag tag-name)
      (org-roam-node-list))))
(defun my/org-roam-filter-by-tags (wanted unwanted)
  (lambda (node)
  (let ((node-tags (org-roam-node-tags node)))
    (and (cl-some (lambda (tag) (member tag node-tags)) wanted)
         (not (cl-some (lambda (tag) (member tag node-tags)) unwanted))))))
(defun my/org-roam-refresh-agenda-list ()
  (interactive)
  (setq org-agenda-files
        (delete-dups (append org-agenda-files
                             (my/org-roam-list-notes-by-tag "Project")))))
;; for projects
(defun my/org-roam-project-finalize-hook ()
  "adds the captured project file to `org-agenda-files' if the
  capture was not aborted."
  ;; remove the hook since it was addd temporarily
  (remove-hook 'org-capture-after-finalize-hook #'my/org-roam-project-finalize-hook)
  ;; add project file to the agenda list if the capture was confirmed
  (unless org-note-abort
    (with-current-buffer (org-capture-get :buffer)
      (add-to-list 'org-agenda-files (buffer-file-name)))))

(defun my/org-roam-insert-new-project ()
  (interactive)
  ;; add the project file to the agenda after capture is finished
  (add-hook 'org-capture-after-finalize-hook #'my/org-roam-project-finalize-hook)
  ;; select a project file to open, creating it if necessary
  (org-roam-capture- :node (org-roam-node-read
                            nil
                            (my/org-roam-filter-by-tag "Project"))
                     :templates (list my/org-roam-project-template)))

(defun my/org-roam-find-project ()
  (interactive)
  ;; add the project file to the agenda after capture is finished
  (add-hook 'org-capture-after-finalize-hook #'my/org-roam-project-finalize-hook)
  ;; select a project file to open, creating it if necessary
  (org-roam-node-find
   nil
   nil
   (my/org-roam-filter-by-tags '("Project") '("Archived"))))
;; new todo in project
(defun my/org-roam-capture-task ()
(interactive)
;; update org-agenda list after adding projects
(add-hook 'org-capture-after-finalize-hook #'my/org-roam-project-finalize-hook)
;; new todo
(org-roam-capture- :node (org-roam-node-read
                          nil
                          (my/org-roam-filter-by-tag "Project"))
                   :templates (list my/org-roam-project-template)))
(defun my/org-roam-copy-todo-to-today ()
 (interactive)
 (unless (or (string= (buffer-name) "*habit*") ; do nothing in habit buffer
         (string= (org-entry-get nil "STYLE") "habit")) ; skip if the task is a habit
   (let ((org-refile-keep t) ; set this to nil to delete the original!
         (org-roam-dailies-capture-templates
          '(("t" "tasks" entry "%?"
             :if-new (file+head+olp "%<%Y-%m-%d>.org" "#+title: %<%Y-%d-%d>\n" ("Done")))))
         (org-after-refile-insert-hook #'save-buffer)
         today-file
         pos)
     (save-window-excursion
       (org-roam-dailies--capture (current-time) t)
       (setq today-file (buffer-file-name))
       (setq pos (point)))
     ;; only refile if the target file is different than the current file
     (unless (equal (file-truename today-file)
                    (file-truename (buffer-file-name)))
       (org-refile nil nil (list "Done" today-file nil pos))))))

;; Org-roam configuration for Doom Emacs
(after! org-roam
  (setq org-roam-directory "~/Documents/org/notes/"
        org-roam-completion-everywhere t
        org-roam-node-display-template
        (concat "${title:*} " (propertize "${tags:10}" 'face 'org-tag))
        org-roam-db-gc-threshold most-positive-fixnum
        org-roam-dailies-directory "daily/"
        org-roam-dailies-capture-templates
        `(("d" "default" entry "* %?"
            :if-new (file+head ,my/daily-note-filename
                              ,my/daily-note-header)))
        org-roam-capture-templates
        '(("h" "Hugo Blog Post" plain
          (file "~/Documents/org/templates/hugo-post.org") ;; Uses external template
            :target (file+head "%<%y%m%d%h%m%s>-${slug}.org" "")
            :unnarrowed t)))
  ;; Keybindings
  (map! :leader
        :desc "Toggle org-roam buffer" "n r l" #'org-roam-buffer-toggle
        :desc "Capture org-roam note" "n r c" #'org-roam-capture
        :desc "Insert org-roam node" "n r i" #'org-roam-node-insert
        :desc "Insert immediate org-roam node" "n r I" #'org-roam-node-insert-immediate
        :desc "Capture org-roam task" "n r t" #'my/org-roam-capture-task
        :desc "Insert new project" "n r P" #'my/org-roam-insert-new-project
        :desc "Find project" "n r p" #'my/org-roam-find-project
        :desc "Toggle org-roam UI" "n r u" #'org-roam-ui-mode)

  ;; Additional keybindings for Org mode
  (map! :map org-mode-map
        "C-M-i" #'completion-at-point)
  (add-hook! 'after-init-hook #'my/org-roam-refresh-agenda-list))

(use-package! org-roam-ui
  :config
    (setq org-roam-ui-sync-theme t)
    (setq org-roam-ui-follow t)
    (setq org-roam-ui-update-on-save t)
    (setq org-roam-ui-open-on-start t))

(org-babel-do-load-languages
  'org-babel-load-languages
  '((emacs-lisp . t)
    (plantuml . t)
    (python . t)))
(setq org-confirm-babel-evaluate nil)
(add-hook 'org-babel-after-execute-hook 'org-redisplay-inline-images)

(after! eee
  (setq ee-terminal-command "st") ; Set terminal command
  (map! :leader
        (:prefix ("t" . "toggle")  ; Prefix for toggle-related commands
         :desc "Lazygit" "g" #'ee-lazygit
         :desc "Yazi" "y" #'ee-yazi)))

(use-package! aidermacs
  :bind (("C-c a" . aidermacs-transient-menu))
  :custom
  ; See the Configuration section below
  (aidermacs-use-architect-mode t)
  (setq aidermacs-backend 'vterm)
  (setq aidermacs-vterm-multiline-newline-key "S-<return>")
  (setq aidermacs-config-file "~/.aider.conf.yml"))

(use-package! gptel
  :config
  ;; Retrieve API key securely
  ;; (setq gptel-api-key (auth-source-pick-first-password :host "api.openai.com" :user "apikey"))
  (setq gptel-api-key
        (auth-source-pick-first-password :host "openrouter.ai" :user "openrouter-apikey"))

  ;; Use OpenRouter's API endpoint
  (setq gptel-backend (gptel-make-openai "OpenRouter"
                      :host "openrouter.ai"
                      :endpoint "/api/v1/chat/completions"
                      :key gptel-api-key
                      :models '(deepseek/deepseek-r1:free
                                openai/gpt-4o-2024-11-20
                                anthropic/claude-3.7-sonnet)))
  ;; Keybinding to quickly open `gptel`
  (map! :leader
     :desc "Chat with GPT via OpenRouter"
     "o p" #'gptel))

(use-package! yasnippet
  :init
  (add-hook 'yas-minor-mode-hook (lambda()
				       (yas-activate-extra-mode 'fundamental-mode)))
  :config
  (setq yas-snippet-dirs '("~/dotconfig/emacs/doom/snippets")))

(after! rime
  (setq rime-user-data-dir "~/dotconfig/rime")
  (setq default-input-method "rime"
        rime-show-candidate 'posframe
        rime-disable-predicates
        '(rime-predicate-evil-mode-p
          rime-predicate-after-ascii-char-p
          rime-predicate-hydra-p
          rime-predicate-tex-math-or-command-p
          rime-predicate-prog-in-code-p))
  ;; Prevent rime crash on exit
  (defun rime-lib-finalize () nil)
  (add-hook 'kill-emacs-hook #'rime-lib-finalize))

(after! pangu-spacing
  (setq pangu-spacing-real-insert-separator t) ;; Enable real spacing
  (global-pangu-spacing-mode 1)) ;; Enable globally
